fax/loop.py

```diff
@@ -151,3 +149,65 @@ def scan_step(args, idx):
         iterations=iterations,
         previous_value=prev_sol,
     )
+
+
+def _debug_fixed_point_iteration(init_x, func, convergence_test, max_iter, batched_iter_size=1,
```

Comment:
Unclear if this should go in this file. @gehring would know better. Also @NeilGirdhar proposed a scan-based idea which may encompass this. Overall, I suggest putting all changes to loop.py in a separate PR since unrelated to extragradient

Reply:
reverting these changes, I should have removed the viz code.

Reply:
In case you're curious, here's [my version](https://github.com/NeilGirdhar/fax_suggestions/blob/master/cmm/fixed_point/iterated_function.py#L81).   Using scan is much faster since it compiles the loop.   The tap function is a new JAX feature that lets us have a progress bar that operates during the scan.  The extract function lets us get information out of the scan.  Also, I have the parameters as a regular object rather than a closure, which allows the parameters to contain tracers and be compiled.  Without this design, you cannot nest this call in another compiled function.

Reply:
@NeilGirdhar As far as I understand, `jax.lax.while_loop` should be compiled in the same situations as `jax.lax.scan`. Scan can possibly have better memory allocation but I'm not sure how well XLA leverages that. From the `jax.lax.scan` docs, "also unlike that Python version, scan is a JAX primitive and is lowered to a single XLA While HLO."

In what setting have you encountered significantly different performance?

Reply:
@gehring I guess you're responding to my statement about "scan being much faster".  I was comparing using scan to Manuel's version that uses a plain Python for loop to sample a trajectory.  I guess that wasn't clear.

Reply:
Ah, my bad! Then yes, I agree.

Though, when jitted, the speed up will mostly be in compile time. In fact, I believe there are some cases where a statically compiled python loop will be faster than scan due to how stuff gets dispatched to the gpu (similar to inefficiencies with while loops having to check with the cpu at every iteration). I recall reading some open issue about it so presumably the XLA team will be improving this at some point.

Small aside, this overhead with GPU + loops is why I added the batching of iterations to the implementation of the fixed point iteration loop.

Reply:
Interesting!